### PR TITLE
[Merged by Bors] - feat(CategoryTheory/Localization): main results on left calculus of fractions

### PR DESCRIPTION
**Pull request by joelriou**
In this PR, given a class of morphisms W : MorphismProperty C which admits a left calculus of fractions, we obtain that if `L : C ⥤ D` is a localization functor for W, then morphisms in `D` can be computed using left fractions, and we show that the functor `L.mapArrow : Arrow C ⥤ Arrow D` is essentially surjective.

---
- [x] depends on: #10606

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/CategoryTheory/Localization/CalculusOfFractions.lean [hunk 1]
```diff
@@ -643,12 +644,129 @@ instance : (Q W).IsLocalization W :=
 
 end
 
+lemma homMk_eq {X Y : C} (f : LeftFraction W X Y) :
+    homMk f = f.map (Q W) (Localization.inverts _ W) := by
+  have := Localization.inverts (Q W) W f.s f.hs
+  erw [← Q_map_comp_Qinv f.f f.s f.hs]
+  rw [← cancel_mono ((Q W).map f.s),
+    assoc, Qiso_inv_hom_id, comp_id, map_comp_map_s]
+
```
Comment (TwoFX) [inline Mathlib/CategoryTheory/Localization/CalculusOfFractions.lean:653]:
```suggestion
```

File: Mathlib/CategoryTheory/Localization/CalculusOfFractions.lean [hunk 2]
```diff
@@ -643,12 +644,129 @@ instance : (Q W).IsLocalization W :=
 
 end
 
+lemma homMk_eq {X Y : C} (f : LeftFraction W X Y) :
+    homMk f = f.map (Q W) (Localization.inverts _ W) := by
+  have := Localization.inverts (Q W) W f.s f.hs
+  erw [← Q_map_comp_Qinv f.f f.s f.hs]
+  rw [← cancel_mono ((Q W).map f.s),
```
Comment (TwoFX) [inline Mathlib/CategoryTheory/Localization/CalculusOfFractions.lean:651]:
```suggestion
  rw [← Q_map_comp_Qinv f.f f.s f.hs, ← cancel_mono ((Q W).map f.s),
```

File: Mathlib/CategoryTheory/Localization/CalculusOfFractions.lean [hunk 3]
```diff
@@ -643,12 +644,129 @@ instance : (Q W).IsLocalization W :=
...
+      (Localization.compUniqFunctor L₁ L₂ W).hom.app X ≫
+        φ.map L₂ (Localization.inverts L₂ W) ≫
+        (Localization.compUniqFunctor L₁ L₂ W).inv.app Y := by
+  let e := Localization.compUniqFunctor L₁ L₂ W
+  have := Localization.inverts L₂ W φ.s φ.hs
+  rw [← cancel_mono (e.hom.app Y), assoc, assoc, e.inv_hom_id_app, comp_id,
+    ← cancel_mono (L₂.map φ.s), assoc, assoc, map_comp_map_s, ← e.hom.naturality]
+  dsimp
+  rw [← Functor.map_comp_assoc, map_comp_map_s]
+  exact e.hom.naturality φ.f
```
Comment (TwoFX) [inline Mathlib/CategoryTheory/Localization/CalculusOfFractions.lean:689]: Not sure if it's really better, but
```suggestion
  simpa [← Functor.map_comp_assoc, map_comp_map_s] using e.hom.naturality φ.f
```
removes the non-terminal `dsimp`.

**Review by TwoFX — APPROVED** [review]
Thanks!
bors d+

**Comment by joelriou** [comment]
Thanks!

bors merge

**Pull request closed without merging**